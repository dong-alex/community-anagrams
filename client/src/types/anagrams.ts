export type Anagram = {
  firstWord: string;
  secondWord: string;
  result: boolean;
};

export type TopAnagram = {
  firstWord: string;
  secondWord: string;
  count: number;
};

export type AnagramReducerState = {
  anagrams: Anagram[];
  topAnagrams: TopAnagram[];
  loading: boolean;
};

export type TopAnagramsBoardProps = {
  topAnagrams: TopAnagram[];
  loading: boolean;
};

export type AnagramSubmission = {
  firstWord: string;
  secondWord: string;
};

export type AnagramRequestCardProps = {
  onSaveAnagram: (
    firstWord: string,
    secondWord: string
  ) => Promise<boolean | never>;
};

export type TopEntryItemProps = {
  index: number;
  firstWord: string;
  secondWord: string;
  count: number;
};
