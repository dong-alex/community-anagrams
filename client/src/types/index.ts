export * from "./anagrams";
export * from "./action";
