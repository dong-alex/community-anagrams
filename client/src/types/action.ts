import { Anagram, TopAnagram } from "./anagrams";

export type AnagramAction = {
  type: string;
  payload?: {
    topAnagrams?: TopAnagram[];
    anagrams?: Anagram[];
    submittedAnagram?: Anagram;
  };
};
