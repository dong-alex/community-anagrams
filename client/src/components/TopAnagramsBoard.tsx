import React, { FunctionComponent } from "react";
import {
  MDBBox,
  MDBListGroup,
  MDBListGroupItem,
} from "mdbreact";
import { TopAnagram, TopAnagramsBoardProps } from "../types";

const TopAnagramsBoard: FunctionComponent<TopAnagramsBoardProps> = ({
  topAnagrams,
}) => {
  return (
    <MDBBox className="mb-5">
      <h5
        className="text-center pt-5 pb-5"
        data-testid="top-anagrams-board-header"
      >
        Top 10 Most Requested Anagram Checks
      </h5>
      <MDBListGroup>
        {topAnagrams.map(
          ({ firstWord, secondWord, count }: TopAnagram, i: number) => (
            <MDBListGroupItem
              data-testid="top-anagram-item"
              key={i}
              className="text-center"
              style={{ userSelect: "none" }}
            >
              {i + 1}. {firstWord} - {secondWord} - {count}
            </MDBListGroupItem>
          )
        )}
      </MDBListGroup>
    </MDBBox>
  );
};

export default TopAnagramsBoard;
