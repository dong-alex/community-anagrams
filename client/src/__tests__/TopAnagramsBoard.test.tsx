import React from "react";
import { render, screen, act } from "@testing-library/react";
import TopAnagramsBoard from "../components/TopAnagramsBoard";
import { TopAnagram } from "../types";

describe("TopAnagramsBoard Tests", () => {
  it("renders the basic template without data", () => {
    const data: TopAnagram[] = [];

    const { container } = render(<TopAnagramsBoard topAnagrams={data} />);

    expect(container).toMatchSnapshot();
  });

  it("renders the list of top anagrams", () => {
    const data: TopAnagram[] = [
      {
        firstWord: "hello",
        secondWord: "world",
        count: 10000,
      },
      {
        firstWord: "aaa",
        secondWord: "world",
        count: 5512,
      },
      {
        firstWord: "test",
        secondWord: "test",
        count: 2221,
      },
      {
        firstWord: "abc",
        secondWord: "abc",
        count: 667,
      },
      {
        firstWord: "www",
        secondWord: "www",
        count: 445,
      },
      {
        firstWord: "ca",
        secondWord: "com",
        count: 323,
      },
      {
        firstWord: "aaaaaaaaaaaaaaaaaaaaaaaaaaa",
        secondWord: "aaaaaaaaaaaaaaaaaaaaaaa",
        count: 100,
      },
      {
        firstWord: "aaaaaaaaaaaaaaaa",
        secondWord: "wwwwwwwwwwwwwwwwwwwwwwww",
        count: 6,
      },
      {
        firstWord: "longestwordiwouldbeabletofit",
        secondWord: "someotherlongwordiwouldbeabletoFit",
        count: 3,
      },
      {
        firstWord: "shortword",
        secondWord: "anothershortWord",
        count: 1,
      },
    ];

    const { getAllByTestId } = render(<TopAnagramsBoard topAnagrams={data} />);

    const anagrams = getAllByTestId("top-anagram-item");
    expect(anagrams).toHaveLength(10);
  });
});
