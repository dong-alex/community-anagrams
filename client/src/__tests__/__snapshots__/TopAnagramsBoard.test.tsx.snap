// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopAnagramsBoard Tests renders the basic template without data 1`] = `
<div>
  <div
    class="mb-5"
    data-test="box"
  >
    <h5
      class="text-center pt-5 pb-5"
      data-testid="top-anagrams-board-header"
    >
      Top 10 Most Requested Anagram Checks
    </h5>
    <ul
      class="list-group"
      data-test="list-group"
    />
  </div>
</div>
`;
